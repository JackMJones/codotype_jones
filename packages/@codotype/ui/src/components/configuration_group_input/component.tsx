import * as React from "react";
import {
    OptionType,
    OptionValue,
    OptionValueInstance,
    ConfigurationGroupProperty,
    ConfigurationGroup,
    GroupLayoutVariant,
} from "@codotype/types";
import { ConfigurationInputChild } from "./ConfigurationInputChild";
import { ConfigurationGroupVariant } from "./ConfigurationGroupVariant";
import { ConfigurationGroupHeader } from "./ConfigurationGroupHeader";
import { ConfigurationInputFormGroup } from "./ConfigurationInputFormGroup";
import { ConfigurationInstanceInput } from "./ConfigurationInstanceInput";
import { ConfigurationCollectionInput } from "./ConfigurationCollectionInput";

// // // //

interface ConfigurationInputProps {
    value: OptionValueInstance;
    configurationGroup: ConfigurationGroup;
    onChange: (updatedVal: OptionValueInstance) => void;
}
export function ConfigurationInput(props: ConfigurationInputProps) {
    if (!props.configurationGroup.properties) {
        console.log("WARNING - NO CONFIGURATION GROUP PROPERTIES DEFINED");
        return null;
    }
    return (
        <div className="row mt-3">
            <div className="col-lg-12">
                {/* ConfigurationGroupHeader */}
                <ConfigurationGroupHeader
                    configurationGroup={props.configurationGroup}
                    enableDocumentationModal={
                        props.configurationGroup.layoutVariant !==
                        GroupLayoutVariant.DOCS_4x8
                    }
                />

                {/* Renders ConfigurationGroup variants */}
                {/* QUESTION - maybe rename to `layout`? */}
                <ConfigurationGroupVariant
                    configurationGroup={props.configurationGroup}
                >
                    {props.configurationGroup.properties.map(
                        (property: ConfigurationGroupProperty) => {
                            if (property.type === OptionType.COLLECTION) {
                                return (
                                    <ConfigurationCollectionInput
                                        label={property.label}
                                        properties={property.properties}
                                        onChange={(updatedVal: OptionValue) => {
                                            props.onChange({
                                                ...props.value,
                                                // [property.identifier]: updatedVal,
                                            });
                                        }}
                                        value={props.value}
                                    />
                                );
                            }

                            // Handle instance input
                            if (property.type === OptionType.INSTANCE) {
                                const val =
                                    // @ts-ignore
                                    props.value[property.identifier];
                                return (
                                    <div
                                        className="card card-body mb-4 mt-2"
                                        key={property.identifier}
                                    >
                                        <ConfigurationInstanceInput
                                            label={property.label}
                                            properties={property.properties}
                                            onChange={(
                                                updatedVal: OptionValueInstance,
                                            ) => {
                                                if (property.allowDisable) {
                                                    props.onChange({
                                                        ...props.value,
                                                        [property.identifier]: {
                                                            enabled: true,
                                                            value: updatedVal,
                                                        },
                                                    });
                                                } else {
                                                    props.onChange({
                                                        ...props.value,
                                                        [property.identifier]: updatedVal,
                                                    });
                                                }
                                            }}
                                            value={val}
                                        />
                                        {/* <p>{property.identifier}</p> */}
                                        {/* <p>{JSON.stringify(property)}</p> */}
                                        {/* Debugging */}
                                        {/* {JSON.stringify(val)} */}
                                    </div>
                                );
                            }

                            // Handle all simple inputs
                            // @ts-ignore
                            const value = props.value[property.identifier];
                            return (
                                <ConfigurationInputFormGroup
                                    card
                                    property={property}
                                    key={property.identifier}
                                >
                                    <ConfigurationInputChild
                                        value={value}
                                        property={property}
                                        onChange={(
                                            updatedValue: OptionValue,
                                        ) => {
                                            props.onChange({
                                                ...props.value,
                                                [property.identifier]: updatedValue,
                                            });
                                        }}
                                    />
                                    {/* Debugging */}
                                    {/* {JSON.stringify(props.value)} */}
                                </ConfigurationInputFormGroup>
                            );
                        },
                    )}
                </ConfigurationGroupVariant>
            </div>
        </div>
    );
}
