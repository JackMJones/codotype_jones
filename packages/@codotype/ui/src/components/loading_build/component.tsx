import * as React from "react";
import { LoadingSpinner } from "../loading_spinner";
const sample = require("lodash.sample");

// // // //

export const quotes: [string, string][] = [
    ["Patience is bitter, but its fruit is sweet.", "Aristotle"],
    ["I have no patience with dinosaurs.", "Adam West"],
    [
        "Have patience with all things, But, first of all with yourself.",
        "Saint Francis de Sales",
    ],
    [
        "Rivers know this: there is no hurry. We shall get there some day.",
        "A.A. Milne",
    ],
    ["Trees that are slow to grow bear the best fruit.", "Moliere"],
    ["Patience is a conquering virtue.", "Geoffrey Chaucer"],
    [
        "The strongest of all warriors are these two â€” Time and Patience.",
        "Leo Tolstoy",
    ],
];

/**
 * LoadingBuild
 * Renders a spinning Codotype Logo
 */
export function LoadingBuild(props: { quote?: [string, string] }) {
    const quote: [string, string] = props.quote || sample(quotes) || quotes[0];
    const quoteBody = quote[0];
    const quoteAuthor = quote[1];

    return (
        <div className="h-100 d-flex flex-row justify-content-center align-items-center">
            <div className="col-sm-12 col-md-8 col-lg-6">
                <div className="card card-body text-center shadow-sm">
                    <h3>
                        codotype{" "}
                        <span
                            style={{
                                fontWeight: "lighter",
                            }}
                        >
                            is generating your codebase
                        </span>
                    </h3>
                    <p className="text-muted mb-0">
                        This should only take a few moments
                        <i className="far fa-laugh" />
                    </p>
                    <div className="row my-4">
                        <div className="col-lg-12 text-center d-flex justify-content-center">
                            <div className="my-2">
                                <LoadingSpinner />
                            </div>
                        </div>
                    </div>
                    <small className="text-muted">
                        {quoteBody}
                        <br />- <strong>{quoteAuthor}</strong>
                    </small>
                </div>
            </div>
        </div>
    );
}
