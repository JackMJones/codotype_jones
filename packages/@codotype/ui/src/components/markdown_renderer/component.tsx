import * as React from "react";
// import ReactMarkdown from "react-markdown";
// import SyntaxHighlighter from "react-syntax-highlighter";

// // // //

// export function CodeBlock(props: { value: string; language: string }) {
//     const { language = undefined, value } = props;
//     return <SyntaxHighlighter language={language}>{value}</SyntaxHighlighter>;
// }

// // // //

// export function ImageBlock(props: { alt: string; src: string }) {
//     return (
//         <img
//             src={props.src}
//             alt={props.alt}
//             style={{
//                 maxWidth: "50%",
//             }}
//         />
//     );
// }

// // // //

/**
 * MarkdownRenderer
 * Renders markdown
 */
export function MarkdownRenderer(props: { source: string }) {
    return (
        <div>
            <pre className="bg-dark text-light px-3 py-3 rounded">
                {props.source}
            </pre>
            {/* <ReactMarkdown
                skipHtml
                source={props.source}
                renderers={{
                    code: CodeBlock,
                    image: ImageBlock,
                }}
            /> */}
        </div>
    );
}
