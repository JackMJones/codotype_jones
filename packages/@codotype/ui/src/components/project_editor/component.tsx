import React, { FunctionComponent } from "react";
import { ConfigurationGroupSelector } from "./ConfigurationGroupSelector";
import { ProjectEditorHeader } from "./ProjectEditorHeader";
import { GeneratorMeta, Project } from "@codotype/types";

// // // //

interface ProjectEditorProps {
    generator: GeneratorMeta;
    project: Project;
    onChange: (updatedProject: Project) => void;
    onClickGenerate: () => void;
    onResetProject: () => void;
    // enableImport?: boolean;
    // enableExport?: boolean;
}

export const ProjectEditor: FunctionComponent<ProjectEditorProps> = (
    props: ProjectEditorProps,
) => {
    const { generator, project } = props;
    return (
        <div className="card card-body shadow-sm">
            <div className="row">
                <div className="col-sm-12">
                    {/* Render ProjectEditorHeader */}
                    <ProjectEditorHeader
                        generatorMeta={generator}
                        project={project}
                        onChange={props.onChange}
                        onClickGenerate={props.onClickGenerate}
                        onConfirmReset={props.onResetProject}
                    />

                    {/* Render ConfigurationGroupSelector */}
                    <ConfigurationGroupSelector
                        project={project}
                        generatorMeta={generator}
                        onChange={(updatedProject: Project) => {
                            // Invokes props.onChange with the updated project
                            console.log("UPDATED PROJECT");

                            props.onChange(updatedProject);
                        }}
                    />
                </div>
            </div>
        </div>
    );
};
