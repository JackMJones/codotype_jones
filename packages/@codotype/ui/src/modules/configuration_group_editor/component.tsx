import * as React from "react";
import { string } from "prop-types";

// // // //

/**
 * MoreInfoLinkProps
 * Props for the `MoreInfoLink` component
 * @param url - the URL thats linked to
 */
interface MoreInfoLinkProps {
  url: string;
}

/**
 * MoreInfoLink
 * Renders a link that when clicked opens a new tab to some relevant information
 */
export function MoreInfoLink(props: MoreInfoLinkProps) {
  return (
    <a href={props.url} target="_blank">
      More Info
    </a>
  );
}

// // // //
// // // //
// // // //

// Example OptionValueInstance
// {
//   componentName: "MoreInfoLink",
//     componentSlug: "more_info_link",
//       props: [
//         { type: "number", name: "age", desc: "desc" }
//     { type: "function", name: "onChange", desc: "desc" }
//       ],
//         tests: {
//     enabled: true,
//       properties: {
//       type: "TABLE_TEST"
//     }
//   }
// }

export enum OptionType {
  STRING = "STRING",
  NUMBER = "NUMBER",
  BOOLEAN = "BOOLEAN",
  DROPDOWN = "DROPDOWN",
  MULTI_DROPDOWN = "MULTI_DROPDOWN",
  COLLECTION = "COLLECTION",
  INSTANCE = "INSTANCE",
}

export type OptionValue =
  | string
  | string[]
  | number
  | number[]
  | boolean
  | OptionValueInstance
  | OptionValueInstance[]
  | null;

type OptionValueInstance =
  | OptionValueInstanceStandard
  | OptionValueInstanceAllowDisable;

interface OptionValueInstanceStandard {
  [key: string]: OptionValue;
}

interface OptionValueInstanceAllowDisable {
  enabled: boolean;
  value: OptionValue;
}

interface DropdownOption {
  label: string;
  value: string;
}

interface ConfigurationGroupProperty {
  label: string;
  identifier: string;
  description: string;
  type: OptionType;
  defaultValue: OptionValue;
  required: boolean;
  enabled: boolean;
  allowDisable: boolean;
  properties: ConfigurationGroupProperty[];
  dropdownOptions: DropdownOption[];
}

interface ConfigurationGroup {
  label: string;
  identifier: string;
  description: string;
  enabled: boolean;
  allowDisable: boolean;
  scope: string; // enum
  variant: string; //enum
  properties: ConfigurationGroupProperty[];
}

// ^^^^ this is built from:

export const ComponentBuilderConfigurationGroupProperty: ConfigurationGroupProperty = {
  label: "Component",
  identifier: "components",
  description: "Define the component data",
  type: OptionType.COLLECTION,
  defaultValue: [],
  enabled: true,
  required: false,
  allowDisable: false,
  dropdownOptions: [],
  properties: [
    {
      label: "Component Name",
      identifier: "componentName",
      description: "Name of the component",
      type: OptionType.STRING,
      defaultValue: "",
      enabled: true,
      required: false,
      allowDisable: false,
      properties: [],
      dropdownOptions: [],
    },
    {
      label: "Component Slug",
      identifier: "componentSlug",
      description: "Slug of the component",
      type: OptionType.STRING,
      defaultValue: "",
      enabled: true,
      required: false,
      allowDisable: false,
      properties: [],
      dropdownOptions: [],
    },
    {
      label: "Props",
      identifier: "props",
      description: "",
      dropdownOptions: [],
      type: OptionType.COLLECTION,
      defaultValue: "",
      enabled: true,
      required: false,
      allowDisable: false,
      properties: [
        {
          label: "Type",
          identifier: "type",
          description: "",
          type: OptionType.DROPDOWN,
          defaultValue: "",
          enabled: true,
          required: false,
          allowDisable: false,
          dropdownOptions: [{ label: "String", value: "string" }],
          properties: [],
        },
        {
          label: "Name",
          identifier: "name",
          description: "",
          type: OptionType.STRING,
          defaultValue: "",
          enabled: true,
          required: false,
          allowDisable: false,
          dropdownOptions: [],
          properties: [],
        },
        {
          label: "Desc",
          identifier: "desc",
          description: "",
          type: OptionType.STRING,
          defaultValue: "",
          enabled: true,
          required: false,
          allowDisable: false,
          dropdownOptions: [],
          properties: [],
        },
      ],
    },
    {
      label: "Tests",
      identifier: "tests",
      description: "",
      type: OptionType.INSTANCE,
      defaultValue: "",
      enabled: true,
      required: false,
      allowDisable: true,
      dropdownOptions: [],
      properties: [
        {
          label: "Test Type",
          identifier: "testType",
          description: "",
          type: OptionType.DROPDOWN,
          defaultValue: "",
          enabled: true,
          required: false,
          allowDisable: false,
          dropdownOptions: [],
          properties: [],
        },
      ],
    },
  ],
};

export const ComponentBuilderConfigurationGroup: ConfigurationGroup = {
  label: "Component Generator",
  identifier: "components_group",
  description: "Generate React components",
  enabled: true,
  allowDisable: false,
  variant: "",
  scope: "",
  properties: [ComponentBuilderConfigurationGroupProperty],
};

/**
 * buildConfigurationGroupPropertyValue
 * Accepts a single ConfigurationGroupProperty and recursively produces its associated OptionValue instance
 * @param property - the single ConfigurationGroupProperty for which the OptionValue is being produced
 */
export function buildConfigurationGroupPropertyValue(
  property: ConfigurationGroupProperty
): OptionValue {
  if (property.type === OptionType.STRING) {
    return property.defaultValue || "";
  }
  if (property.type === OptionType.BOOLEAN) {
    return property.defaultValue || false;
  }
  if (property.type === OptionType.DROPDOWN) {
    return property.defaultValue || "";
  }
  if (property.type === OptionType.COLLECTION) {
    return [buildConfigurationGroupValue(property.properties)];
    // Should return this:
    // return [];
  }
  if (property.type === OptionType.INSTANCE) {
    return buildConfigurationGroupValue(property.properties);
  }
  return "";
}

/**
 * buildConfigurationGroupValue
 * Builds the top-level OptionValueInstance for a single ConfigurationGroup
 * @param properties - array of ConfigurationGroupProperty
 */
export function buildConfigurationGroupValue(
  properties: ConfigurationGroupProperty[]
): OptionValueInstance {
  const initialValue: OptionValueInstance = {};

  // Defines empty ConfigurationGroupValue
  // Iterates over each property in the group and assigns values
  const configurationGroupValue: any = properties.reduce(
    (val, property: ConfigurationGroupProperty) => {
      // Updates val with data for ConfigurationGroupProperty
      if (property.allowDisable && !property.required) {
        val[property.identifier] = {
          enabled: property.enabled,
          value: buildConfigurationGroupPropertyValue(property),
        };
      } else {
        val[property.identifier] = buildConfigurationGroupPropertyValue(
          property
        );
      }
      // Returns val
      return val;
    },
    initialValue // Passes in empty OptionValueInstance
  );

  // Returns ConfigurationGroupValue
  return configurationGroupValue;
}

// // // //
