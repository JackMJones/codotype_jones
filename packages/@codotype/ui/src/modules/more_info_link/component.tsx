import * as React from "react";
import { string } from "prop-types";

// // // //

/**
 * MoreInfoLinkProps
 * Props for the `MoreInfoLink` component
 * @param url - the URL thats linked to
 */
interface MoreInfoLinkProps {
  url: string;
}

/**
 * MoreInfoLink
 * Renders a link that when clicked opens a new tab to some relevant information
 */
export function MoreInfoLink(props: MoreInfoLinkProps) {
  return (
    <a href={props.url} target="_blank">
      More Info
    </a>
  );
}
