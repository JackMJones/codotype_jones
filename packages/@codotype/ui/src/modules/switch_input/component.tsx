import * as React from "react";

// // // //

/**
 * SwitchInputProps
 * Props for the `SwitchInput` component
 * @param value - the current value of the switch (true/false)
 * @param onChange - callback function to update `props.value`
 */
interface SwitchInputProps {
  value: boolean;
  onChange: (updatedValue: boolean) => void;
}

/**
 * SwitchInput
 * Renders an `<input type="checkbox" />` element
 */
export function SwitchInput(props: SwitchInputProps) {
  return (
    <input
      type="checkbox"
      checked={props.value}
      onChange={e => {
        props.onChange(e.currentTarget.checked);
      }}
    />
  );
}
