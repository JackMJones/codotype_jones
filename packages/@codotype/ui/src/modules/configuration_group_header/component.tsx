import * as React from "react";
import { MoreInfoLink } from "../more_info_link";

// // // //

/**
 * ConfigurationGroupHeaderProps
 * Props for the `ConfigurationGroupHeader` component
 * @param title - the title in the header
 * @param description - the description in the header
 * @param moreInfoUrl - (optional) the URL passed into `MoreInfoLink`
 */
interface ConfigurationGroupHeaderProps {
  title: string;
  description: string;
  moreInfoUrl?: string;
}

/**
 * ConfigurationGroupHeader
 * Renders the header for a single ConfigurationGroup
 */
export function ConfigurationGroupHeader(props: ConfigurationGroupHeaderProps) {
  return (
    <div>
      <p>{props.title}</p>
      {props.moreInfoUrl && <MoreInfoLink url={props.moreInfoUrl} />}
      <p>{props.description}</p>
    </div>
  );
}
