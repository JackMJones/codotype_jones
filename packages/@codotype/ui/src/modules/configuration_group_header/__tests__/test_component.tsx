import * as React from "react";
import TestRenderer from "react-test-renderer";
import { ConfigurationGroupHeader } from "../component";

describe("ConfigurationGroupHeader", () => {
  test("render", () => {
    const tree = TestRenderer.create(
      <ConfigurationGroupHeader
        title={"My Configuration Group"}
        description={"My Configuration Group Description"}
        moreInfoUrl={"https://codotype.org"}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
