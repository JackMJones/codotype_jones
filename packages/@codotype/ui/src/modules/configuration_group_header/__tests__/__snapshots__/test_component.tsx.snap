// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigurationGroupHeader render 1`] = `
<div>
  <p>
    My Configuration Group
  </p>
  <a
    href="https://codotype.org"
    target="_blank"
  >
    More Info
  </a>
  <p>
    My Configuration Group Description
  </p>
</div>
`;
